src/handler/mod.rs

```diff
@@ -893,51 +914,100 @@ impl Handler {
-                            // this.
-                        }
-                        HandlerReqId::External(id) => {
-                            if let Err(e) = self
-                                .service_send
-                                .send(HandlerOut::RequestFailed(id, request_error))
-                                .await
-                            {
-                                warn!("Failed to inform that request failed {}", e);
-                            }
+                warn!("Failed to send next pending request {request_error}");
+                // Inform the service that the request failed
+                match req.request_id {
+                    HandlerReqId::Internal(_) => {
+                        // An internal request could not be sent. For now we do nothing about
+                        // this.
+                    }
+                    HandlerReqId::External(id) => {
+                        if let Err(e) = self
+                            .service_send
+                            .send(HandlerOut::RequestFailed(id, request_error))
+                            .await
+                        {
+                            warn!("Failed to inform that request failed {e}");
                         }
                     }
                 }
             }
         }
     }
 
+    /// Replays all active requests for the given node address, in the case that a new session has
+    /// been established. If an optional message nonce is provided, the corresponding request will
+    /// be skipped, eg. the request that established the new session.
+    async fn replay_active_requests<P: ProtocolIdentity>(
+        &mut self,
+        node_address: &NodeAddress,
+        // Optional message nonce to filter out the request used to establish the session.
+        message_nonce: Option<MessageNonce>,
+    ) {
+        trace!(
+            "Replaying active requests. {}, {:?}",
+            node_address,
+            message_nonce
+        );
+
+        let packets = if let Some(session) = self.sessions.get_mut(node_address) {
+            let mut packets = vec![];
+            for request_call in self
+                .active_requests
+                .get(node_address)
+                .unwrap_or(&vec![])
+                .iter()
+                .filter(|req| {
+                    // Except the active request that was used to establish the new session, as it has
+                    // already been handled and shouldn't be replayed.
+                    if let Some(nonce) = message_nonce.as_ref() {
+                        req.packet().message_nonce() != nonce
+                    } else {
+                        true
+                    }
+                })
+            {
+                let new_packet = session
+                    .encrypt_message::<P>(self.node_id, &request_call.encode())
+                    .unwrap();
```

Comment:
I kinda believe this but we don't really use `unwrap` in prod code (release builds). Can you please change this to handle the error or at least change it to an `expect` with the right explanation?

This is my only requirement before merging

Reply:
@njgheorghita - I had a quick look through the AES library we use. From what I can tell this function can fail if the size of the messages are over a specific limit. I was looking at version 0.9. I haven't looked at future versions. 

I don't think it's reasonable to assume that in the current version and all possible future versions, that every possible message we try to encrypt will be successful. If there is an event that it's not successful, I dont think we should crash the entire application. 

Failing to encrypt a message is not a catastrophic failure imo. I think it's fine to just log a CRIT message and continue running. 

In which case we shouldn't use an `unwrap()` or `expect()` rather check the failure and if it fails log a crit. 

What do you think?

Reply:
Some thoughts...

- We're only dealing with requests retrieved from the `active_requests` store. IIUC this implies a pretty strong guarantee that the encryption will be successful, since the request has already been successfully encrypted when it originated. If this is accurate, then `expect` might actually expose an error in the request-handling logic - though, I'm unsure whether or not that is worth crashing...
- If the encryption fails, do we want to do more than just log a CRIT message? Maybe it's worth removing the request entirely from the `active_requests` store. Although... maybe it's alright to just leave the (unable to encrypt) active request as is and let it simply timeout? 

> Failing to encrypt a message is not a catastrophic failure imo. 

Ultimately, I agree with this. So I'm leaning towards logging a message. The question I would need to think more about (it's been a while since I've been intimate with this code 😅 ) is whether or not we need to handle un-encryptable requests? Or if it's ok to just log the message and let the request timeout

Reply:
The reason why I more or less believed this could be true is because we are replaying. In that scenario, the appropriate message in `expect` is *why* this is infallible, not simply the error occurred. Maybe could have been more clear about it.

But it seems to be replaying might no be enough guarantee from @AgeManning 's comment? In that case the error needs to be handled. Here that potentially means bobbling up the error to the application. You should look into how failed requests and failed encryption in particular is handled in other parts of the code to get a good idea of what to do in this scenario.

Also, yes to logging the error (there is no CRIT, just ERROR)

> Or if it's ok to just log the message and let the request timeout

Would need to check but I assume we try to guarantee that if a request times out then it is active, removing from one place and not the other would most likely break that invariant.